Enable throttler in activator

vvraskin opened the pull request:
<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #2381

## Proposed Changes

* Enable throttler in `activator/main.go` and in `pkg/activator/handler/handler.go`.
* Define endpoint observer and revision observers in the main method.
* Return 503 in case of throttler overflow (should not happen unless the BreakersQueueDepth is reached), and 502 in case if the revision is not found (should not happen)
* See the reduction of previously observed 503s - https://github.com/knative/serving/issues/1846#issuecomment-459338147
* Extend existing unit tests

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```
Throttle requests in activator in case of overflow.
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -64,6 +71,16 @@ const (
 	// Add enough buffer to not block request serving on stats collection
 	requestCountingQueueLength = 100
+
+	// The number of requests that are queued on the breaker before the 503s are sent
```
```suggestion
	// The number of requests that are queued on the breaker before the 503s are sent.
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -64,6 +71,16 @@ const (
 	requestCountingQueueLength = 100
+
+	// The number of requests that are queued on the breaker before the 503s are sent
+	// the value must be adjusted depending on the actual production requirements
```
```suggestion
	// The value must be adjusted depending on the actual production requirements.
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -64,6 +71,16 @@ const (
+	// the value must be adjusted depending on the actual production requirements
+	breakerQueueDepth = 10000
+
+	// The upper bound for concurrent requests sent to the revision,
```
```suggestion
	// The upper bound for concurrent requests sent to the revision.
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -64,6 +71,16 @@ const (
+	breakerQueueDepth = 10000
+
+	// The upper bound for concurrent requests sent to the revision,
+	// as new endpoints show up, the Breakers concurrency increases up to this value
```
```suggestion
	// As new endpoints show up, the Breakers concurrency increases up to this value.
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -64,6 +71,16 @@ const (
+	// as new endpoints show up, the Breakers concurrency increases up to this value
+	breakerMaxConcurrency = 1000
+
+	defaultResyncInterval = 10 * time.Hour
```
@mattmoor should we factor the default we set (in main.go of the controller) out to a common place to reuse?

mattmoor replied:
Maybe a constant in `pkg/reconciler` ot even `pkg/controller` in `knative/pkg` would make sense.

vvraskin replied:
could add it in a separate PR - to keep this one from growing even more (already XL 😞).

mattmoor replied:
SGTM

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,84 @@ func main() {
+			// No endpoints exist yet.
+			return 0, nil
+		}
+		if exists && endpoints != nil {
```
The exists check is redundant here. Should we rather do a similar check like in the `revisionGetter' below?

```golang
endpoints, ok := endpoints.(*corev1.Endpoints)
if !ok {
  return 0, fmt.Errorf("...")
}

addresses := activator.EndpointsAddressCount(endpoints.Subsets)
return int32(addresses), nil
```

That'd keep the "good" path flat and the error handlings in the indented ifs.

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,84 @@ func main() {
+
+	// Return the revision from the observer.
+	// Additionally to observer errors an error is returned if either no revision exists yet
+	// or we could not cast the interface to Revision for some reason.
```
I'd keep this narrowed to the first line. If you want more, the sentence should probably be:

```
An error is returned if either the revision doesn't exist yet or we could not cast to a Revision.
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,84 @@ func main() {
+	// or we could not cast the interface to Revision for some reason.
+	revisionGetter := func(revID activator.RevisionID) (revision *v1alpha12.Revision, err error) {
+		revisionKey := cache.ExplicitKey(revID.Namespace + "/" + revID.Name)
+		rev, ok, err := revisionInformer.Informer().GetIndexer().Get(revisionKey)
```
These two lines are equivalent to:

```golang
revision, err := revisionInformer.Lister().Revisions(revID.Namespace).Get(revID.Name)
```

Same above for endpoints. Don't be thrown off by this being called a Lister. It gives it a Lister interface but it's backed by the Indexer.

If you need to check for not-found, you can do so by importing `apierrors "k8s.io/apimachinery/pkg/api/errors"` and then do `if apierrors.IsNotFound(err) {`

vvraskin replied:
Got it, then we could delegate the existence checking to the lister. Nice to cut some lines.

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,84 @@ func main() {
+		}
+		if !ok {
+			return nil, fmt.Errorf("Error getting revision for %v", revID)
+		}
```
Do we want to return an error for `!ok` here? If so, we could be more explicit and state that the revision does not exist.

vvraskin replied:
I'm going to replace it with the lister, the error should come from it.

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,84 @@ func main() {
+	throttler := activator.NewThrottler(throttlerParams)
+
+	// Update/create the breaker in the throttler when the number of endpoints changes.
+	endpointInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Do we need to listen for "Add" here as well?

vvraskin replied:
As far as I remember the endpoint is getting added with 0 value and then it is updated to 1. So we don't need to watch for the add.

markusthoemmes replied:
I'd say let's be better safe then sorry. We have utility functions to transform an `Add` handler into an `Update` handler.

In particular, your UpdateHandler already ignores the first argument we can straight ahead drop it from the signature and change this to this:

```golang
endpointInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
	AddFunc:    activator.UpdateEndpoints(throttler),
	UpdateFunc: controller.PassNew(activator.UpdateEndpoints(throttler)),
})
```

WDYT?

vvraskin replied:
Makes sense, was not aware of the PassNew trick.

markusthoemmes commented on pkg/activator/handler/handler.go:
```diff
@@ -34,19 +33,36 @@ type ActivationHandler struct {
+	}
+}
+
+func (a *ActivationHandler) sendRequest(w http.ResponseWriter, r *http.Request, namespace, name string, start time.Time, ar activator.ActivationResult) {
```
Should this be called "proxyRequest"?

Can we get away with only passing w and r in here? We could easily handle the timing and the reporting in the outer function. WDYT?

markusthoemmes commented on pkg/activator/handler/handler.go:
```diff
@@ -34,19 +33,36 @@ type ActivationHandler struct {
+	revID := activator.RevisionID{namespace, name}
+
+	// ActiveEndpoint() will block until the first endpoint is available.
+	ar := a.Activator.ActiveEndpoint(namespace, name)
```
Let's handle the error of the ActivationRequest in this function rather than passing it down?

vvraskin commented:
Regarding the tests. 
I actually see that they are bit mixed up with the current handler tests (since the throttling logic is clenched in the retrying handler at the moment). I'd like to refactor them in the upcoming PRs though. Where I'm planning to replace the revision polling with the Informer. This should allow to split the Throttling handler from the Retrying logic/Metric pushing handler. And it should be way easier to test the throttler without the overhead of the rest of the activator.

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +148,66 @@ func main() {
+	// Return the number of endpoints, 0 if no endpoints are found.
+	endpointsGetter := func(revID activator.RevisionID) (int32, error) {
+		endpoints, err := endpointInformer.Lister().Endpoints(revID.Namespace).Get(revID.Name)
+		if err != nil {
```
This used to have error handling for when endpoints cannot be found to default to 0. Should we keep that?

You can achieve that using:

```golang
if errors.IsNotFound(err) {
  return 0, nil
}
```

vvraskin replied:
ah, thanks you, deleted it in the rush of happily doing the replacement with the Lister :)

markusthoemmes commented on pkg/activator/handler/handler.go:
```diff
@@ -34,30 +33,52 @@ type ActivationHandler struct {
+		if err == activator.ErrActivatorOverload {
+			http.Error(w, activator.ErrActivatorOverload.Error(), http.StatusServiceUnavailable)
+		} else {
+			w.WriteHeader(http.StatusInternalServerError)
```
Let's log this non overload error.

vagababov commented on pkg/activator/handler/handler_test.go:
```diff
@@ -268,6 +308,159 @@ func TestActivationHandler(t *testing.T) {
 
+// Make sure we return http internal server error when the Breaker is overflowed
+func TestActivationHandler_Overflow(t *testing.T) {
+	wantedSuccess := 20
```
those are consts?

vagababov replied:
Same below.

vvraskin replied:
The var should be fine in this case I think, since we do not reuse the same value. Unless you have a strong reason to make it a const.

vagababov replied:
they don't have to be global consts, ie. 
```
func... {
const (
a = b
c = d
)
...
}
```

vvraskin replied:
Sure, if you feel that they would make the code better, I can add them.

vagababov replied:
Constants are constants :) it's more expressive that way.

vagababov commented on pkg/activator/handler/handler_test.go:
```diff
@@ -268,6 +308,159 @@ func TestActivationHandler(t *testing.T) {
+		if err != nil {
+			return resp, err
+		}
+		time.Sleep(50 * time.Millisecond)
```
Why is there a sleep here?

vvraskin replied:
When we test sending 21 requests we have only 20 available slots, if the requests are too fast, the 21st could get a free slot from one of the "speedy" requests. Thus the sleep to make sure the last request is always rejected. 
I added a comment.

vagababov replied:
Can we do this in a channel or a wait group of sorts? Sleeps usually generate lots of flakes.

vvraskin replied:
Do you have an exact place where you would have placed them?

Please consider my comment above, I'm planning to refactor the tests in the next PRs. It would be way easier to do that once the throttler is separated from the activating handler.

vagababov replied:
If you're going to refactor them -- sure, we can leave it until later.

vagababov commented on pkg/activator/handler/handler_test.go:
```diff
@@ -268,6 +308,159 @@ func TestActivationHandler(t *testing.T) {
+		case successCode:
+			gotBody, _ := ioutil.ReadAll(resp.Body)
+			if string(gotBody) != wantBody {
+				t.Errorf("Unexpected response body. Want %q, got %q", wantBody, gotBody)
```
got before want is the preferred way. (i.e.g ResponseBody = %q, want: %q)

vvraskin replied:
I wouldn't say it is consistent across all the tests and it makes a big difference, but I could change it if you think it is necessary.

vagababov replied:
I'd really appreciate. And I unify that wherever I see it.
Thank you!

vvraskin commented:
The previous tests failed because of "Scale to 50" - will be rerun after the push. There was one failure of "BuildPipelineAndServe" which looked like an istio/envoy blip with "no healthy upstream".

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+	endpointInformer := kubeInformerFactory.Core().V1().Endpoints()
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
```
```suggestion
	kubeInformerFactory.Start(stopCh)
```

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
+	go revisionInformer.Informer().Run(stopCh)
```
```suggestion
	servingInformerFactory.Start(stopCh)
```

mattmoor commented on cmd/activator/main.go:
```diff
@@ -39,14 +39,21 @@ import (
 	"github.com/knative/serving/pkg/activator"
 	activatorhandler "github.com/knative/serving/pkg/activator/handler"
 	activatorutil "github.com/knative/serving/pkg/activator/util"
+	v1alpha12 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
```
if you are going to alias this, I'd go with `servingv1alpha1`.

vvraskin replied:
I dropped the alias.

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+
+	// Return the revision from the observer.
+	revisionGetter := func(revID activator.RevisionID) (*v1alpha12.Revision, error) {
+		rev, err := revisionInformer.Lister().Revisions(revID.Namespace).Get(revID.Name)
```
```suggestion
		return revisionInformer.Lister().Revisions(revID.Namespace).Get(revID.Name)
```

mattmoor replied:
...and delete the trailing lines.

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+		return rev, err
+	}
+
+	throttlerParams := activator.ThrottlerParams{BreakerParams: params, Logger: logger, GetEndpoints: endpointsGetter, GetRevision: revisionGetter}
```
```suggestion
	throttlerParams := activator.ThrottlerParams{
		BreakerParams: params,
		Logger:        logger,
		GetEndpoints:  endpointsGetter,
		GetRevision:   revisionGetter,
	}
```

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+	}
+
+	throttlerParams := activator.ThrottlerParams{BreakerParams: params, Logger: logger, GetEndpoints: endpointsGetter, GetRevision: revisionGetter}
+	throttler := activator.NewThrottler(throttlerParams)
```
So the throttler takes into consideration the concurrency of the target revision, but I wonder if/how it takes into consideration that the activator itself may be horizontally scaled?

If the activator deployment has `N` pods, then we should be throttling the traffic we send to each Pod of the Revision to `{concurrency} / N`, vs. just `{concurrency}`.

Maybe you are doing this and I just haven't gotten there yet, but I wanted to make sure I wrote it down.

vvraskin replied:
At the moment we don't do it. Would be happy to send a separate PR to enable this. Right now we do have only a single activator instance (there are > 1 in the tests through), I think it would be enough for the scope of initial throttling enablement.

mattmoor replied:
I'm fine with separating it, but wanted to confirm this is a problem (and also have it factor into your thinking as you chase this).  Can you open an issue to track and add it to the project dashboard (or DM me a link and I can?).

thanks.

vvraskin replied:
Here is the issue: https://github.com/knative/serving/issues/3165

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+	})
+
+	kubeInformerFactory.Start(stopCh)
+	servingInformerFactory.Start(stopCh)
```
🤔 Ok, so with these, you can simply delete the lines above where I suggested you replace them with these...

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+
+	// Update/create the breaker in the throttler when the number of endpoints changes.
+	endpointInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    activator.UpdateEndpoints(throttler),
```
We should probably add a Filter on this, so that we only consider endpoints that have labels indicating they were transitively created by a Revision.

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +150,70 @@ func main() {
+
+	// Remove the breaker if the revision was deleted.
+	revisionInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: activator.DeleteBreaker(throttler),
```
Why not simply do this on Endpoints as well?

vvraskin replied:
I'm not sure whether we need to do that, if an endpoint is deleted it could simply mean that we scaled down to 0 and we would need the breaker once new requests come in.

mattmoor replied:
I believe the Endpoints resource is deleted iff the K8s Service is deleted, and we don't do that unless the Revision is deleted.

vvraskin replied:
Got it, will update it to delete the breaker on the Endpoints.

mattmoor commented on pkg/activator/handler/handler_test.go:
```diff
@@ -82,23 +102,25 @@ func TestActivationHandler(t *testing.T) {
-			wantCode:  http.StatusOK,
-			wantErr:   nil,
-			attempts:  "123",
+			label:           "active endpoint",
```
nit: you can dramatically reduce indentation in these cases by using:
```go
{{
   ...
}. {
  ...
}. {
  ...
}}
```

For the array literal.

mattmoor commented on pkg/activator/throttler_test.go:
```diff
@@ -190,8 +190,8 @@ func TestThrottler_TryOverload(t *testing.T) {
-	if got := err.Error(); got != OverloadMessage {
-		t.Errorf("Error message = %q, want: %q", got, OverloadMessage)
+	if got := err; got != ErrActivatorOverload {
+		t.Errorf("Error message = %q, want: %q", got, ErrActivatorOverload.Error())
```
```suggestion
		t.Errorf("Error message = %v, want: %v", got, ErrActivatorOverload)
```

mattmoor commented on pkg/activator/handler/handler_test.go:
```diff
@@ -29,8 +30,21 @@ import (
 	. "github.com/knative/pkg/logging/testing"
 	"github.com/knative/serving/pkg/activator"
 	"github.com/knative/serving/pkg/activator/util"
+	v1alpha12 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
```
`servingv1alpha1`

vvraskin replied:
Dropped the alias.

mattmoor commented on pkg/activator/handler/handler.go:
```diff
@@ -34,30 +33,53 @@ type ActivationHandler struct {
+	})
+	if err != nil {
+		if err == activator.ErrActivatorOverload {
+			http.Error(w, activator.ErrActivatorOverload.Error(), http.StatusServiceUnavailable)
```
I'm wondering if this should be using a 429?

vvraskin replied:
Generally agree, but I think it would lead to inconsistency with the 503s we send back from the queue-proxy overflow. So I prefer to leave it with 503 for now, we could replace it 429 once we redirect the queue-proxy overflow back to the activator.

mattmoor replied:
Can you open an issue to discuss the use of 429 vs. 5xx?

It is sort of cosmetic, but I think 429 is a much clearer status code (at least to me).  503 and 504 seem to grow on trees :)

markusthoemmes replied:
I vaguely recall discussing this somewhere. The bottomline was:

- If your system runs out of resources (as in this case), return 503.
- If the user exhausted his quotas, use 429.

vvraskin replied:
Got it, that depends on whether we consider the size of a breaker to be a user limit or not. Technically if the user is running out of throttler tokens, we might still have the capacity to provision containers. 

I opened this issue:
https://github.com/knative/serving/issues/3166

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -133,6 +152,81 @@ func main() {
+		revisionID := serving.RevisionUID
+		_, ok := endpoints.Labels[revisionID]
+		return ok
+	}
```
Can we generalize this in `pkg/reconciler/filter.go` as an `AnnotationExistsFilterFunc`? There's already a filter function there to check for specific annotations. This seems useful (and possibly allows us to inline `handler` and that function into the `AddEventHandler` call below?

vvraskin commented:
@markusthoemmes 
Yeap, I'm planning to add an e2e tests after this one gets in. 
On a side note, the latest unit tests/build failures due to a single left-over method, I cleaned it up. 

@tcnghia could you please review/approve? I touched the reconciler files. 
@mattmoor I moved the DeleteFunc to the Endpoints informer.

markusthoemmes commented on pkg/reconciler/filter.go:
```diff
@@ -34,3 +35,12 @@ func AnnotationFilterFunc(key string, value string, allowUnset bool) func(interf
+// LabelExistsFilterFunc creates a FilterFunc only accepting objects which have a given label.
+func LabelExistsFilterFunc(label string) func(obj interface{}) bool {
+	return func(obj interface{}) bool {
+		endpoints := obj.(*corev1.Endpoints)
```
This is not generalized. This function can operate on any metav1.Object (see the function above), like

```golang
func LabelExistsFilterFunc(label string) func(obj interface{}) bool {
	return func(obj interface{}) bool {
		if mo, ok := obj.(metav1.Object); ok {
			labels := mo.GetLabels()
			_, ok := labels[label]
			return ok
		}
		return false
	}
}
```

vvraskin commented:
The preUpgrade test seems to be flaky in the pipeline, and it passes locally, so 
/test pull-knative-serving-upgrade-tests

vvraskin commented:
scaleTo50 flake...
/test pull-knative-serving-integration-tests

vvraskin commented:
hmr, unrelated test failures (their path doesn't go through the activator)

/test pull-knative-serving-integration-tests

vvraskin commented:
Unrelated unit test failure, 
/test pull-knative-serving-unit-tests

markusthoemmes commented:
/assign @mattmoor 

For the leftover approval.

mattmoor commented on cmd/activator/main.go:
```diff
@@ -133,6 +152,73 @@ func main() {
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go revisionInformer.Informer().Run(stopCh)
+	go endpointInformer.Informer().Run(stopCh)
```
I think you removed the wrong lines :)

These two should start the informer factories, which don't need `go` in front.

vvraskin replied:
As per slack discussion, I added a comment about that. 
In short, I intentionally Run the informers instead of starting them from the factory. Otherwise, they are hanging waiting for a sync.
